- Scikit learn is the machine learning library of Python 
- There are also a few (7) in built datasets which are very helpful to begin with understanding the concepts and also to build a quick POC.

In [7]:
import sklearn
print(sklearn.__version__)

0.23.1


**how to load a csv file that has a space delimiter and that exists without column names?**

In [11]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 
                'PTRATIO', 'B', 'LSTAT', 'MEDV']


In [12]:
boston_csv_df = pd.read_csv(r"D:\SimpliLearn\Machine+Learning\boston_dataset\housing.csv",
                            header=None, delimiter=r"\s+", names=column_names)

In [13]:
boston_csv_df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [16]:
boston_csv_df.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV'],
      dtype='object')

**how to load a csv file?**

In [20]:
import pandas as pd

In [34]:
boston_csv_df2 = pd.read_csv(r"C:\Users\Vaishali Balaji\Desktop\ML_Notes\HousingData.csv")

In [19]:
boston_csv_df2.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [23]:
boston_csv_df2.dtypes

CRIM       float64
ZN         float64
INDUS      float64
CHAS       float64
NOX        float64
RM         float64
AGE        float64
DIS        float64
RAD          int64
TAX          int64
PTRATIO    float64
B          float64
LSTAT      float64
MEDV       float64
dtype: object

In [27]:
boston_csv_df2.shape

(506, 14)

In [47]:
boston_csv_df2.isnull().sum()

CRIM       20
ZN         20
INDUS      20
CHAS       20
NOX         0
RM          0
AGE        20
DIS         0
RAD         0
TAX         0
PTRATIO     0
B           0
LSTAT      20
MEDV        0
dtype: int64

In [49]:
boston_csv_df2["CRIM"].fillna(boston_csv_df2["CRIM"].mean() , inplace = True)

In [50]:
boston_csv_df2["ZN"].fillna(boston_csv_df2["ZN"].mean() , inplace = True)

In [51]:
boston_csv_df2["INDUS"].fillna(boston_csv_df2["INDUS"].mean() , inplace = True)

In [52]:
boston_csv_df2["CHAS"].fillna(boston_csv_df2["CHAS"].mean() , inplace = True)

In [53]:
boston_csv_df2["AGE"].fillna(boston_csv_df2["AGE"].mean() , inplace = True)
boston_csv_df2["LSTAT"].fillna(boston_csv_df2["LSTAT"].mean() , inplace = True)

In [54]:
boston_csv_df2.isnull().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

From the dataset, we conclude the following:
- There are 506 rows and 14 columns in total.
- MEDV column represents the target variable.

# Split data into X & y

In [55]:
X = boston_csv_df2.drop("MEDV", axis=1) #mapping only the features to the X variable
y = boston_csv_df2["MEDV"] #mapping the target column to y variable

In [56]:
X.shape

(506, 13)

In [57]:
y.shape

(506,)

# Split data into train and test set

In [58]:
from sklearn.model_selection import train_test_split

In [67]:
X_train, X_test, y_train,y_test = train_test_split(X,y,test_size=0.2, random_state=21)

test_size is a parameter tht can accept any value between 0 and 1. But mind that 0 and 1 are exclude (i.e) it can accept any value>0 and <1

for ex, if test_size = 0.2, it implies that 20% of the data points are to be used for testing and the remaining 80% of the data points are to be used for training.

for ex, if test_size = 0.1, it implies that 10% of the data points are to be used for testing and the remaining 90% of the data points are to be used for training.

- test_size = 0 ==> None of the data points to be used for testing
- test_size = 1 ==> 100% of data to be used for testing  ==> this is the reason for the test_size parameter to never accept the values 0 and 1.


In [68]:
print("Shape of train set features")
print(X_train.shape)
print("Shape of test set features")
print(X_test.shape)

Shape of train set features
(404, 13)
Shape of test set features
(102, 13)


In [69]:
print("Shape of train set target")
print(y_train.shape)
print("Shape of test set target")
print(y_test.shape)

Shape of train set target
(404,)
Shape of test set target
(102,)


# Building the Multiple Linear Regression model
- Step 1 : Import a model of our choice.
- Step 2 : Instantiate an estimator object which contains the ML algorithm.
- Step 3 : Fit the ML algorithm on the given data (train set)
- Step 4 : Use the generated model on the test set for predictions.
- Step 5 : Evaluate your model using appropriate evaluation metrics.

In [70]:
from sklearn.linear_model import LinearRegression

In [71]:
model_lr = LinearRegression()

In [72]:
model_lr.fit(X_train,y_train) #supervised machine learning

LinearRegression()

In [73]:
print("The intercept for this Multiple Linear regression model on the boston dataset is ", 
      model_lr.intercept_)

The intercept for this Multiple Linear regression model on the boston dataset is  38.907779177974916


In [74]:
print("The regression co-efficients of the Multiple Linear regression model are :")
print("==========================================================================")
print(model_lr.coef_)

The regression co-efficients of the Multiple Linear regression model are :
[-9.28601688e-02  4.38263969e-02 -3.41389762e-02  3.50417947e+00
 -1.70604743e+01  3.58509226e+00 -1.25968368e-02 -1.49138985e+00
  2.87996826e-01 -1.02797819e-02 -1.02796179e+00  1.07407706e-02
 -4.90868188e-01]


In [75]:
y_test_predicted_values = model_lr.predict(X_test)
#Always only the test data features should be passed as input to the predict method

In [76]:
y_test_predicted_values

array([15.30305359, 15.99912087, 27.00066968, 37.42812356, 33.36966308,
       12.53967536, 25.68908198, 30.12669218, 33.30972184, 18.23344367,
       22.44137553, 16.42105539, 28.74804553,  9.43073569, 18.83742095,
       28.10990227, 16.05446918, 29.15571818,  4.18986325, 17.94624014,
       20.91539482, 26.37358837, 19.51738773,  2.11146863, 20.88680461,
       34.31312197, 25.47853414, 31.56067572, 13.31686845, 25.08957433,
       25.62424663, 18.73795594, 34.12095465, 13.286765  , 24.29245345,
       31.33389701, 31.45192514, 21.83113765, 20.98614641, 17.49366501,
       19.72949717, 32.9967124 , 24.45376012, 24.50457769, 22.24138382,
       20.61348996, 30.09729841, 21.2695684 , 30.68697434, 23.35050504,
       34.45568844, 14.66472984, 41.10736943, 16.39265602, 13.70410158,
       11.52988752,  8.21762413, 23.55944391, 22.67043755, 27.24493692,
       17.02500618, 35.20013668, 25.56520325, 22.56826434, 36.22793317,
       36.43051212, 28.39851872, 19.46571006, 21.5708352 , 19.53

In [77]:
y_test_predicted_values.shape

(102,)

In [79]:
import pandas as pd
pd.DataFrame({'Actual y_test': y_test, 'Predicted y_test':y_test_predicted_values})

Actual y_test  Predicted y_test
455           14.1         15.303054
142           13.4         15.999121
311           22.1         27.000670
232           41.7         37.428124
290           28.5         33.369663
..             ...               ...
486           19.1         19.606019
468           19.1         17.323622
302           26.4         29.278697
244           17.6         16.378167
321           23.1         24.863574

[102 rows x 2 columns]

# Evaluating the MLR model

In [80]:
print("R2 score is:{}".format(model_lr.score(X_test,y_test)))

R2 score is:0.689039418687388


In [81]:
from sklearn.metrics import mean_squared_error
import numpy as np

In [82]:
print("RMSE value of the test dataset is :")
print(np.sqrt(mean_squared_error(y_test, y_test_predicted_values))) #(actual values,predicted values)

RMSE value of the test dataset is :
5.409472374503402


# End of discussion